ng-dev/release/notes/changelog.ts

```diff
@@ -0,0 +1,141 @@
+ */
+const joinMarker = `\n\n${splitMarker}\n\n`;
+
+/** A RegExp matcher to extract the version of a changelog entry from the entry content. */
+const versionAnchorMatcher = new RegExp(`<a name="(.*)"></a>`);
+
+/** An individual changelog entry. */
+interface ChangelogEntry {
+  content: string;
+  version: semver.SemVer;
+}
+
+export class Changelog {
+  /** The absolute path to the changelog file. */
+  readonly filePath = join(this.git.baseDir, changelogPath);
+  /** The absolute path to the changelog archive file. */
+  readonly archiveFilePath = join(this.git.baseDir, changelogArchivePath);
+
+  /** The changelog entries in the CHANGELOG.md file. */
+  private entries = this.getEntriesFor(this.filePath);
+  /**
+   * The changelog entries in the CHANGELOG_ARCHIVE.md file.
+   * Delays reading the CHANGELOG_ARCHIVE.md file until it is actually used.
+   */
+  private get archiveEntries() {
+    if (this._archiveEntries === undefined) {
+      return (this._archiveEntries = this.getEntriesFor(this.archiveFilePath));
+    }
+    return this._archiveEntries;
+  }
+  private _archiveEntries: undefined | ChangelogEntry[] = undefined;
+  /** Whether the changelog archive has entries upon intialization. */
+  private alreadyHasChangelogArchive = this.archiveEntries.length !== 0;
+
+  constructor(private git: GitClient) {}
+
+  /** Prepend a changelog entry to the changelog. */
+  prependEntryToChangelog(entry: string) {
+    this.entries.unshift(parseChangelogEntry(entry));
+    this.writeToChangelogFiles();
+  }
+
+  /**
+   * Move all changelog entries from the CHANGELOG.md file for versions prior to the provided
+   * version to the changelog archive.
+   *
+   * Versions should be used to determine which entries are moved to archive as versions are the
+   * most accurate piece of context found within a changelog entry to determine its relationship to
+   * other changelog entries.  This allows for example, moving all changelog entries out of the
+   * main changelog when a version moves out of support.
+   */
+  moveEntriesPriorToVersionToArchive(version: semver.SemVer) {
+    [...this.entries].reverse().forEach((entry: ChangelogEntry) => {
+      if (version.compare(entry.version) === 1) {
+        this.archiveEntries.unshift(entry);
+        this.entries.splice(this.entries.indexOf(entry), 1);
+      }
+    });
+    this.writeToChangelogFiles();
+  }
+
+  /** Update the changelog files with the known changelog entries. */
+  private writeToChangelogFiles(): void {
+    const changelog = this.entries.map((entry) => entry.content).join(joinMarker);
+    writeFileSync(this.filePath, changelog);
+
+    // Only update the changelog archive if it previously had content, or if there is now content.
+    if (this.alreadyHasChangelogArchive || this.archiveEntries.length) {
+      const changelogArchive = this.archiveEntries.map((entry) => entry.content).join(joinMarker);
+      writeFileSync(this.archiveFilePath, changelogArchive);
```

Comment:
How about just having a separate method for writing the archive instead? Right now we'd parse the changelog archive file unnecessarily when just an entry has been prepended?